File: backend/app/schemas/printer.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import PrinterDriverType


class PrinterBase(BaseModel):
    name: str
    driver_type: PrinterDriverType
    host: str | None = None
    port: int | None = None
    serial_number: str | None = None
    location: str | None = None


class PrinterCreate(PrinterBase):
    access_code: str | None = None


class PrinterUpdate(BaseModel):
    name: str | None = None
    host: str | None = None
    port: int | None = None
    serial_number: str | None = None
    location: str | None = None
    access_code: str | None = None
    is_active: bool | None = None


class PrinterOut(PrinterBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    is_active: bool
    created_at: datetime
    has_access_code: bool = False


class PrinterStatusOut(BaseModel):
    status: str
    progress: float
    nozzle_actual: float
    nozzle_target: float
    bed_actual: float
    bed_target: float
    current_job_id: uuid.UUID | None = None


class PrinterTestResult(BaseModel):
    success: bool
    message: str

File: frontend/src/lib/printer.ts
export type PrinterDriverType = "BAMBU_LAN" | "OCTOPRINT" | "MOONRAKER" | "DUMMY";

export const driverTypeLabels: Record<PrinterDriverType, string> = {
  BAMBU_LAN: "Bambu Lab (LAN mode)",
  OCTOPRINT: "OctoPrint (unsupported)",
  MOONRAKER: "Moonraker (unsupported)",
  DUMMY: "Dummy (testing/demo)",
};

export interface Printer {
  id: string;
  name: string;
  driver_type: PrinterDriverType;
  host: string | null;
  port: number | null;
  serial_number: string | null;
  location: string | null;
  is_active: boolean;
  has_access_code: boolean;
  created_at: string;
}

export interface PrinterStatus {
  status: string;
  progress: number;
  nozzle_actual: number;
  nozzle_target: number;
  bed_actual: number;
  bed_target: number;
}

export interface PrinterTestResult {
  success: boolean;
  message: string;
}
